maint_tools/check_defaults.py

```diff
@@ -0,0 +1,167 @@
+                    update_desc = False
+                elif i == lineno:
+                    update_def = True
+                    update_desc = False
+
+                if update_def and line.startswith(f"    {param.arg_name} :"):
+                    f.write(f"    {param.arg_name} : {type_name}\n")
+                    print(" updating type name")
+                    update_def = False
+                    update_desc = True
+
+                elif update_def and line.startswith(
+                    f"        {param.arg_name} :"
+                ):
+                    f.write(f"        {param.arg_name} : {type_name}\n")
+                    print(" updating type name")
+                    update_def = False
+                    update_desc = True
+
+                elif update_desc:
+                    if line == f"        {target_str}{default}.\n":
+                        f.write("")
+                        print(" updating description")
+                        update_desc = False
+                    elif line.endswith(f" {target_str}{default}.\n"):
+                        f.write(line.replace(f" {target_str}{default}.", ""))
+                        update_desc = False
+                        print(" updating description")
+                    else:
+                        f.write(line)
+
+                else:
+                    f.write(line)
+
+
+def check_functions(body, file):
+    """Check functions of a module or methods of a class."""
+    for node in body:
+        if isinstance(node, ast.FunctionDef):
+            if SKIP_PRIVATE and node.name.startswith("_"):
+                continue
+            # print(f"function: '{node.name}' in "
+            # "{file.resolve()}:{node.lineno}")
+            docstring = ast.get_docstring(node)
+            docstring = check_docstring(docstring, file, node.lineno)
+
+
+def main():
+    """Update defaults."""
+    modules = (root_dir() / "nilearn").glob("**/*.py")
+
+    files_to_skip = ["test_", "conftest.py"]
+
+    for file in modules:
+        if (
+            any(file.name.startswith(s) for s in files_to_skip)
+            or file.parent.name == "tests"
+        ):
+            continue
+
+        if SKIP_PRIVATE and (
+            file.name.startswith("_") or file.parent.name.startswith("_")
+        ):
+            continue
+
+        with open(file) as f:
+            module = ast.parse(f.read())
+
+        check_functions(module.body, file)
+
+        class_definitions = [
+            node for node in module.body if isinstance(node, ast.ClassDef)
+        ]
+
+        for class_def in class_definitions:
+            # print(
+            #     f"class: '{class_def.name}' "
+            #     f"in {file.resolve()}:{class_def.lineno}"
+            # )
+
+            docstring = ast.get_docstring(class_def)
+            check_docstring(docstring, file, class_def.lineno)
+
+            check_functions(class_def.body, file)
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
Originally I though that this was a script it would be convenient to "keep around" for maintainers to run "once in a while" to ensure that defaults are declared properly.

But after looking around it feels like the one from @mathdugre in the sandbox:

https://github.com/nilearn/nilearn_sandbox/blob/master/nilearn_sandbox/_utils/missing_docstring_default.py

may be adapted to do a better and more systematic job than I put together here for moving defaults.

This should also be discussed in the context of @ymzayek work to mkae type annotation play nice with writing and rendering of the docstrings in https://github.com/nilearn/nilearn/pull/4049

To keep things "simple" I would suggest:
- move this script to the sandbox for now
- modify @mathdugre scripts to make sure that all functions / methods with default parameters have those defaults described in the docstring (at least for the public API)
- cross the bridge about docstring + type annotation when we get to it.

Reply:
move the script to a PR in the sandbox